Setup Base Unit Test Module

rohit9625 opened the pull request:
[KMPPT-17](https://mifosforge.jira.com/browse/KMPPT-17)

Work done in this PR:
- Added base testing module and test dispatchers
- Added test koin modules to provide test-specific implementations:
    - **TestDispatchersModule** for providing `StandardTestDispatchers` and `UnconfinedTestDispatchers` when required
    - **TestDatabaseModule** for providing in-memory database implementation for `nativeTest`, androidInstrumentedTest` and `desktopTest`.
    - **TestCoreDatastoreModule** for providing a test implementation of `UserPreferencesDataStore` by using `MapSettings` instead of `Settings` and a testDispatcher instead of actual IO dispatcher.
    
- Write sample unit tests for **cmp-android** to test distpatchers
- Write unit tests to test common `UserPreferencesDataStore`
- Write unit tests to test `SampleDao` against desktop, android, and native targets.
- Shouldn't be covered in this PR but also write androidTest to test Navigation flow for **Android**

[KMPPT-17]: https://mifosforge.jira.com/browse/KMPPT-17?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

rohit9625 committed:
chore(build-logic): add android library custom gradle plugin

rohit9625 committed:
test: add testing module to the root directory

rohit9625 committed:
test(testing): create test distpatchers for testing

rohit9625 committed:
test(cmp-android): write sample unit test to test dispatchers

niyajali commented on build-logic/convention/build.gradle.kts:
```diff
@@ -56,6 +56,11 @@ gradlePlugin {
         }
 
+        register("androidLibrary") {
+            id = "org.convention.android.library"
```
We no need to create AndroidLibrary plugin for testing..

rohit9625 replied:
Okay, I'll remove that from the project.

niyajali commented on cmp-android/src/test/kotlin/cmp/android/app/ExampleUnitTest.kt:
```diff
@@ -0,0 +1,73 @@
+import org.mifos.testing.di.TestDispatchersModule
+import kotlin.test.assertEquals
+
+class ExampleUnitTest : KoinTest {
```
The `sourceset `should be `commonTest `and all unit testing files should be reside there

rohit9625 replied:
This test is inside cmp-android. So, we'd have androidTest and androidUnitTest source-sets here, I guess.

niyajali replied:
No need such, Unit Test will be done module wise and the whole End-to-end test could be reside in the application module

niyajali replied:
go though the above mentioned links , read those properly and setup kmp `testing ` module and create KMP testing plugin and include other dependencies required and write Unit Test module wise

rohit9625 replied:
That means we'll not need androidTest or androidUnitTest directories?

niyajali reviewed (changes requested):
@rohit9625 I notice you're setting up testing specifically for the Android platform. Since this is a KMP & CMP project, I'd recommend using the Kotlin Multiplatform testing approach instead. This would allow you to test all platform codebases simultaneously, eliminating the need to create and configure separate tests for each platform. I can share some resources to help you get started with this unified testing approach if you're interested.

https://www.kodeco.com/books/kotlin-multiplatform-by-tutorials/v1.0/chapters/8-testing
https://www.jetbrains.com/help/kotlin-multiplatform-dev/multiplatform-run-tests.html
https://medium.com/@vptarasov/unit-testing-in-kotlin-multiplatform-fb3ca7d5c869https://www.jetbrains.com/help/kotlin-multiplatform-dev/compose-test.html
https://touchlab.co/understanding-and-configuring-your-kmm-test-suite
https://medium.com/@rudradave/kotlin-multiplatform-mini-course-testing-and-deployment-strategies-part-4-0bcc12f40294

rohit9625 commented:
> @rohit9625 I notice you're setting up testing specifically for the Android platform. Since this is a KMP & CMP project, I'd recommend using the Kotlin Multiplatform testing approach instead. This would allow you to test all platform codebases simultaneously, eliminating the need to create and configure separate tests for each platform. I can share some resources to help you get started with this unified testing approach if you're interested.
> 

Yes, I realized that yesterday. I was following the `nowinandroid` project to set up testing and couldn't able to make the testing module accessible to other modules so I just copied the `Android Library` plugin from there and it solved the problem. I am working on creating making testing module a KMP module. Since, I no longer need the Android Library plugin so should I remove that?

rohit9625 committed:
test(testing): convert testing module to kmp module

rohit9625 committed:
test(cmp-android): write tests to verify navigation flow

rohit9625 committed:
refactor(testing): remove redundant manifest file from main source-set

rohit9625 committed:
test(testing): setup test module for CoreDataStore and add data

rohit9625 committed:
refactor(testing): add test coroutine scheduler in TestDispatchersModule

rohit9625 committed:
test(core-base): write tests for UserPreferencesDatastore methods

rohit9625 committed:
test(core/database): write tests for SampleDao for possible targets

rohit9625 committed:
test(core/database): create db test module and replace in-test db creation

HekmatullahAmin commented on build-logic/convention/src/main/kotlin/AndroidLibraryConventionPlugin.kt:
```diff
@@ -0,0 +1,33 @@
+import org.gradle.kotlin.dsl.apply
+import org.gradle.kotlin.dsl.configure
+
+class AndroidLibraryConventionPlugin : Plugin<Project> {
```
- Get rid of this. Our purpose is testing just for two plugin creating a whole plugin is not necessary.
- And also we will not need this for testing.

rohit9625 replied:
Ohh, sorry, I forgot to remove this. I added them initially when facing some errors.

HekmatullahAmin commented on core-base/datastore/src/commonTest/kotlin/org/mifos/corebase/datastore/UserPreferencesDatastoreTest.kt:
```diff
@@ -0,0 +1,113 @@
+import kotlin.test.assertTrue
+
+class UserPreferencesDatastoreTest : KoinTest {
+    private val datastore: UserPreferencesDataStore by inject()
```
- for testing DataStore we will use [russhwolf/multiplatform-settings](https://github.com/russhwolf/multiplatform-settings) MapSettings.
- We will not inject the actula UserPreferencesDataStore. If you want to test it so go read that link and see how it will be done

rohit9625 replied:
Well, I created a `TestCoreDatastoreModule` to provide a test implementation of UserPreferenceDataStore because this it accepts:

```kt
/**
 * A data store class for managing user preferences using a settings storage mechanism.
 * This class provides methods to store, retrieve, and manage key-value pairs in a thread-safe manner
 * using coroutines and a specified dispatcher.
 *
 * @property settings The underlying settings storage implementation used to persist preferences.
 * @property dispatcher The [CoroutineDispatcher] used to execute storage operations, ensuring thread safety.
 */
class UserPreferencesDataStore(
    private val settings: Settings,
    private val dispatcher: CoroutineDispatcher,
) {
```

This is the module that passes the MapSettings when creating `UserPreferencesDataStore:
```kt

val TestCoreDataStoreModule = module {
    single<Settings> { MapSettings() }
    single {
        UserPreferencesDataStore(
            get(),
            get(named(AppDispatchers.IO.name)),
        )
    }
}
```

HekmatullahAmin commented on core/database/build.gradle.kts:
```diff
@@ -26,16 +29,33 @@ kotlin {
             implementation(libs.androidx.sqlite.bundled)
         }
 
+        nativeTest.dependencies {
```
- When we say base-unit it means we will use that module as a base for all modules and platforms(ios, android, desktop, web).
- and we will apply the dependency of it in commonTest.
```
commonTest.dependencies {
     implementation(projects.testing)
}
```

rohit9625 replied:
Doest that means, implementing `testing` module in `commonTest` makes it accessible to nativeTest, desktopTest, and androidUnitTest as well?

HekmatullahAmin commented on core/database/src/nativeTest/kotlin/org/mifos/core/database/SampleDaoTest.kt:
```diff
@@ -0,0 +1,58 @@
+import kotlin.test.assertEquals
+
+class SampleDaoTest : KoinTest {
+    private val db: AppDatabase by inject()
```
- note we will not use actual DB. Use mocks, fakes, or in-memory databases.

```
@RunWith(AndroidJUnit4::class)
class UserDaoTest {

    private lateinit var db: AppDatabase
    private lateinit var userDao: UserDao

    @Before
    fun setUp() {
        db = Room.inMemoryDatabaseBuilder(
            ApplicationProvider.getApplicationContext(),
            AppDatabase::class.java
        ).allowMainThreadQueries().build()

        userDao = db.userDao()
    }

    @After
    fun tearDown() {
        db.close()
    }

    @Test
    fun insertAndGetUser() = runBlocking {
        val user = User(id = 1, name = "John")
        userDao.insert(user)
        val fetched = userDao.getById(1)
        assertEquals("John", fetched?.name)
    }
}
```

rohit9625 replied:
Yes, the `TestDatabaseModule` is all about providing inMemoryDatabase for defined targets.

```kt
package org.mifos.testing.di

import org.koin.core.module.Module
import org.koin.dsl.module
import org.mifos.core.database.AppDatabase

val TestDatabaseModule = module {
    includes(testPlatformModule)
    single { get<AppDatabase>().sampleDao }
}

@Suppress("NO_ACTUAL_FOR_EXPECT")
expect val testPlatformModule: Module
```

For example `TestDatabaseModule.desktop.kt
```kt
actual val testPlatformModule: Module = module {
    single {
        Room.inMemoryDatabaseBuilder<AppDatabase>()
            .setDriver(BundledSQLiteDriver())
            .build()
    }
}
```

HekmatullahAmin commented on testing/build.gradle.kts:
```diff
@@ -0,0 +1,54 @@
+
+kotlin {
+    sourceSets {
+        commonMain.dependencies {
```
- We will do unit test module by module. So whatever gets test will be accessible there and will use koinInject for it.
- So all these dependencies that you added for `commonMain`, `androidMain` and etc here they should not be added here.
- We need this dependency block here:
```
commonTest.dependencies {
       implementation(libs.kotlin.test)
       // and other common test
   }
```

rohit9625 replied:
I added them because creating `TestModule` for providing test-specific implementation also requires specific dependencies for each target. However, I am not very familiar with `Gradle` in kmp-projects so correct me if I'm wrong :)

HekmatullahAmin commented on testing/src/commonMain/kotlin/org/mifos/testing/data/UsersTestData.kt:
```diff
@@ -0,0 +1,30 @@
+
+import org.mifos.testing.model.TestUserData
+
+val usersTestData: List<TestUserData> = listOf(
```
Don't need to add sample data here, cause this one will not be used for all modules. whatever resides in this module will be used by whole module

rohit9625 replied:
I think a template contains boilerplate code which often includes such demo tests and test data. I had to add the demo data to test if `DataStore` correctly stores custom data classes as well. 
It is just like a placeholder so that someone writing tests knows that the `testing/data` package contains test data.

HekmatullahAmin commented on testing/src/commonMain/kotlin/org/mifos/testing/di/TestCoreDatastoreModule.kt:
```diff
@@ -0,0 +1,27 @@
+import org.mifos.core.common.di.AppDispatchers
+import org.mifos.corebase.datastore.UserPreferencesDataStore
+
+val TestCoreDataStoreModule = module {
```
instead of creating test modules here we will do it in that specific module and will use DI to pass it to this module. I will provide you what i mean in a comment section of this PR.

rohit9625 replied:
Okay, I'll be happy to discuss this topic. Thanks @HekmatullahAmin for reviewing this PR :)

HekmatullahAmin commented:
Hey @rohit9625 ,

Just wanted to share a few points regarding the recent changes and our approach to testing in the kmp-project-template. The goal behind creating the base-unit module is to ensure common testing logic is accessible across all modules. This module will only include logic that's shared across all projects, and not code that's specific to a single module or project.

Here are a few points to address:

- Please remove the items that Mr. @niyajali  and I previously mentioned.

- We'll not use actual implementations for repositories or databases in tests — instead, we'll use fakes.

- I noticed that in one module you've created commonTest, while in others you've used desktopTest or nativeTest. Some modules may not need platform-specific test folders. Unless Mr. Rajan has explicitly instructed otherwise, for now, just create commonTest directories where necessary.

- Many modules have duplicate setup/teardown code. This kind of boilerplate can be moved into the base-testing module for reuse.

I’ve also been thinking about a more structured approach for core-base testing, and here’s a proposed structure. If @niyajali  and @therajanmaurya  are aligned with this, feel free to move ahead and implement it:

**Proposed Testing Structure**
To simplify and standardize unit testing across the KMP modules, we can introduce a core-base/unittest module:

```
📦 core-base/
├── 📂 unittest/             # Base Module for KMP Unit Tests
│   ├── 📂 src/commonTest/
│   │   ├── 📂 kotlin/
│   │   │   ├── CoroutineTest.kt      # Base class for coroutine-based tests
│   ├── build.gradle.kts
📦 core/
├── 📂 unittest/             # Project-Specific Unit Tests
│   ├── 📂 src/commonTest/
│   │   ├── 📂 kotlin/
│   │   │   ├── ExampleTest.kt        # ViewModel Test using Fake Data
│   │   │   ├── TestModule.kt         # Koin DI Test Module
│   ├── build.gradle.kts

```

**CoroutineTest.kt**
A reusable base class to manage coroutine dispatchers and Koin DI setup:
```
open class CoroutineTest(val testModule: Module) : KoinTest {
    val testDispatcher = StandardTestDispatcher()

    @Before fun setUp() {
        startKoin { modules(testModule) }
        Dispatchers.setMain(testDispatcher)
    }

    @After fun tearDown() {
        stopKoin()
        Dispatchers.resetMain()
    }

    fun runTestWithScope(block: suspend TestScope.() -> Unit) = runTest(testDispatcher) {
        block()
    }
}

```

**TestModule.kt**
An example test DI module using fake implementations:
```
val testModule = module {
    single<UserPreferencesRepository> { FakeUserPreferencesRepository() }
    single { SettingsViewmodel(get()) }
}

```

**ExampleTest.kt**
Shows how to test the ViewModel using the base class and fakes:
```
class ExampleTest : CoroutineTest(testModule = testModule) {
    private val settingsViewmodel: SettingsViewmodel by inject()

    @Test
    fun `test updating theme brand`() = runTestWithScope {
        settingsViewmodel.updateThemeBrand(ThemeBrand.DEFAULT)

        val result = settingsViewmodel.settingsUiState.first { it is SettingsUiState.Success }
        val successResult = result as SettingsUiState.Success

        assertEquals(ThemeBrand.DEFAULT, successResult.settings.brand)
    }
}

```

rohit9625 committed:
Revert "chore(build-logic): add android library custom gradle plugin"

This reverts commit bf02a2c2456c4b5ad00cdca68588db587e1b9428.

rohit9625 commented:
> - We'll not use actual implementations for repositories or databases in tests — instead, we'll use fakes.

I get it, but where I used actual implementations. Please tell me so that I can replace them with fake ones.

> - Many modules have duplicate setup/teardown code. This kind of boilerplate can be moved into the base-testing module for reuse.

Yeah, I agree with that, and that's confusing. I spent lots of time understanding how and where the dependencies should be defined to avoid redundancy. Don't worry; I'll try to avoid duplication.

rohit9625 committed:
refactor(testing): move test koin modules to specific directories

rohit9625 commented:
Hey @HekmatullahAmin. What other changes are required here? I haven't removed the test data from the testing module yet. If I do that, then I need to remove the corresponding test from the `UserPreferencesDatastoreTest`.

HekmatullahAmin committed:
Refactor tests in UserPreferencesDatastoreTest and remove unused code and libraries across files.

HekmatullahAmin commented:
Hi Mr. @therajanmaurya ,

As per the Jira task "Setup Base Unit test module", I and @rohit9625  created the base testing module—currently it's empty, as discussed, and we’ll add common logic to it as we identify reusable components.

Also, I’ve kept the core-base/datastore tests and refactored them following good testing practices, including proper naming conventions (thingUnderTest_triggerOfTest_ResultOfTest) and added KDoc documentation for clarity.

Please let me know if any changes are needed. Thanks!

<img width="1000" alt="Screenshot 2025-04-02 at 19 14 26" src="https://github.com/user-attachments/assets/27054430-c358-4792-9750-c933258d0379" />

rohit9625 commented:
Looks great, thanks @HekmatullahAmin

rohit9625 commented:
Hi @HekmatullahAmin. I just saw you removed 80% of what I did in this PR. I know some of them may be out of the scope of this PR or task. But, they weren't seems to be irrelevant. I know this task was not about writing tests, but I wrote them to verify if we can write tests without any hassle. NavigationTest covers the navigation flow on Android because navigation can be different on Android, IOS, desktop, and web. I thought we would write Navigation tests for other targets in the task `Setting up UI testing`.

Also, what was the reason for removing SampleDaoTest for all targets?

rohit9625 commented:
It's good that you refactored `UserPreferencesDatastoreTest` and added appropriate comments but what was the reason for removing `TestCoreDatastoreModule`? Currently, it's been initialized inside test file itself. What if we are writing tests for a class or ViewModel that requires `UserPreferencesDatastore` then how would be provide that to the test class? The test module was just for making test-specific `UserPreferencesDatastore` accessible to any module implementing `testing` module.

HekmatullahAmin commented:
> Hi @HekmatullahAmin. I just saw you removed 80% of what I did in this PR. I know some of them may be out of the scope of this PR or task. But, they weren't seems to be irrelevant. I know this task was not about writing tests, but I wrote them to verify if we can write tests without any hassle. NavigationTest covers the navigation flow on Android because navigation can be different on Android, IOS, desktop, and web. I thought we would write Navigation tests for other targets in the task `Setting up UI testing`.
> 
> Also, what was the reason for removing SampleDaoTest for all targets?

Hi @rohit9625, I hope you're doing well! I removed the navigation tests because they seemed a bit out of scope for unit testing. I think it would make more sense to address them in UI testing when Rajan assigns that task to you.

As for the SampleDaoTest.kt, I deleted it because we weren't entirely sure if the way we were injecting the Database and Dao was correct. There's no specific documentation for Room KMP testing yet, and the [Android documentation](https://developer.android.com/training/data-storage/room/testing-db) (which is the closest we have) suggests a slightly different approach for injection. Let's definitely have a discussion soon to figure out the best way to implement it.

rohit9625 commented:
> Hi @rohit9625, I hope you're doing well! I removed the navigation tests because they seemed a bit out of scope for unit testing. I think it would make more sense to address them in UI testing when Rajan assigns that task to you.

Okay, I get it. If that has already been done, then it doesn't make sense to remove it. That was the initial work I did in this PR. Someone should've told me that earlier.

> As for the SampleDaoTest.kt, I deleted it because we weren't entirely sure if the way we were injecting the Database and Dao was correct. There's no specific documentation for Room KMP testing yet, and the [Android documentation](https://developer.android.com/training/data-storage/room/testing-db) (which is the closest we have) suggests a slightly different approach for injection. Let's definitely have a discussion soon to figure out the best way to implement it.

Yep, there is no official documentation available yet, and we may need to adapt to some other approach over time, but that doesn't make the current approach wrong. The tests were working fine for all Android, Desktop, and it should also work for iOS as well (if you can run them).
No worries, we'll discuss that in the standup meeting :)

rohit9625 commented:
Well, I just realized that I would've written the SampleDaoTest in the commonTest module only, as it was testing the SampleDao in the commonMain. I forgot that we can run the tests inside commonTest for all targets. And TestDatabaseModule will be responsible for providing their own platform-specific in-memory test database implementation.

I will make the changes if required after the discussion :)

HekmatullahAmin commented:
> It's good that you refactored `UserPreferencesDatastoreTest` and added appropriate comments but what was the reason for removing `TestCoreDatastoreModule`? Currently, it's been initialized inside test file itself. What if we are writing tests for a class or ViewModel that requires `UserPreferencesDatastore` then how would be provide that to the test class? The test module was just for making test-specific `UserPreferencesDatastore` accessible to any module implementing `testing` module.

- I removed TestCoreDatastoreModule from the separate file and initialized it within the test file to ensure the same instance of testDispatcher is used for both UserPreferencesDataStore and runTest, avoiding coroutine context mismatches. If the testDispatcher in the test class and the dispatcher injected into UserPreferencesDataStore were different instances, the runTest block would control virtual time and coroutine execution for its own testDispatcher, but not for the dispatcher used by UserPreferencesDataStore. This could lead to issues like coroutines not executing as expected, or the test completing before UserPreferencesDataStore coroutines finish, it might cause flaky or incorrect results. For example, if runTest(testDispatcher) advances virtual time (e.g., via advanceUntilIdle()), it would only affect coroutines scheduled on testDispatcher, leaving UserPreferencesDataStore coroutines unaffected, potentially causing race conditions or unexecuted coroutines.

- it is a quick approach, for reusability we will think about it. Yeah, maybe as you said we will move it to testing module with a little change
